### PR TITLE
Adding support for cURL in-case allow_url_fopen is 0

### PR DESCRIPTION
**Pull request by nbdamian**
Some servers do not allow requesting https:// when using file_get_contents.
Adding a cURL fallback allows bypassing of this.

### CODE REVIEW

File: src/APIClient.php [hunk 1]
```diff
@@ -307,18 +307,26 @@ protected function call($endpoint, $opts = null) {
...
-                $response_raw = file_get_contents($url, false, $context);
-            }
-        } catch (\Exception $ex) {
-            throw new ApiClientNetworkException($ex->getMessage(), $ex->getCode(), $ex);
-        } finally {
-            restore_error_handler();
-        }
+	    try {
+		    if (function_exists('curl_setopt')) {
+			    $ch = curl_init();
```
Comment (bumperbox) [inline src/APIClient.php]: this takes care of ($opts == null) aka GET request, variation of file_get_contents 
but what about $otps !== null, aka POST request

File: src/APIClient.php [hunk 2]
```diff
@@ -307,18 +307,26 @@ protected function call($endpoint, $opts = null) {
...
+			    $response_raw = curl_exec($ch);
+			    curl_close($ch);
+		    } elseif (ini_get('allow_url_fopen')) {
+			    if ($opts == null) {
+				    $response_raw = file_get_contents($url);
+			    } else {
+				    $context = stream_context_create($opts);
+				    $response_raw = file_get_contents($url, false, $context);
+			    }
+		    }
```
Comment (bumperbox) [inline src/APIClient.php]: } else {
throw new \Exception('File Get Contents or Curl are required');
}

File: src/APIClient.php [hunk 3]
```diff
@@ -307,18 +307,26 @@ protected function call($endpoint, $opts = null) {
...
-            restore_error_handler();
-        }
+	    try {
+		    if (function_exists('curl_setopt')) {
+			    $ch = curl_init();
+			    curl_setopt($ch, CURLOPT_URL, $url);
+			    curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
+			    $response_raw = curl_exec($ch);
+			    curl_close($ch);
+		    } elseif (ini_get('allow_url_fopen')) {
```
Comment (bumperbox) [inline src/APIClient.php]: I would be inclined to put if (ini_get('allow_url_fopen')) { as the first choice, 
then have else if (function_exists('curl_setopt')) { as the fallback option, 
just incase it causes an issue with backwards compatibility

File: src/APIClient.php (hunk 1 above)
Comment (nbdamian) [reply src/APIClient.php]: What do you think of a function in the APIClient that will convert and add the $opts to the cURL instance? or would it be cleaner to add this check outside of the call() function, and maybe supply the $opt in a different format if the server is using cURL??

**Comment by nbdamian** [comment]
> Good initiative, I probably would have gone with curl from the outset myself, but file_get_contents is in that code for historic reasons.
> 
> If you don't want to try and make file_get_contents $opts (POST) work with curl, it looks like it might be a pain to do, we could use just curl or guzzle and just bump the major version and forget about file get contents.
> 
> Your thoughts?

Using only Guzzle would definitely keep the code base cleaner, but would increase the size of the package quite a bit. If this is not a worry then by all means using Guzzle would be good however might be overkill. Perhaps looking into something like Requests for PHP.

**Comment by bumperbox** [comment]
I am happy to go with either, package size shouldn't be a problem

File: src/APIClient.php (hunk 1 above)
Comment (nbdamian) [reply src/APIClient.php]: @bumperbox I have re-written the call() and the build_opts() functions, which should allow for the building of cURL opts or fopen opts. 
https://github.com/whytewaters/api-helper-library/pull/10/commits/9230059632b21ffffc5a1602f47e6be87225f4dd
https://github.com/whytewaters/api-helper-library/pull/10/commits/1b2f3ae38ef22cae7c3452387c0173bdb02fab77

File: src/APIClient.php (hunk 2 above)
Comment (nbdamian) [reply src/APIClient.php]: @bumperbox I have added this Exception as an ApiClientException.
https://github.com/whytewaters/api-helper-library/pull/10/commits/9230059632b21ffffc5a1602f47e6be87225f4dd

File: src/APIClient.php (hunk 3 above)
Comment (nbdamian) [reply src/APIClient.php]: @bumperbox Good idea, I have switched the order around.
https://github.com/whytewaters/api-helper-library/pull/10/commits/9230059632b21ffffc5a1602f47e6be87225f4dd

**Comment by bumperbox** [comment]
@nbdamian thanks or doing those changes, I won't be able to review it util wednesday, will let you know as soon as i have looked

**Comment by nbdamian** [comment]
@bumperbox Have you had some time to have a review this?

**Comment by bumperbox** [comment]
@nbdamian i had a rush trip to hospital, will be back in the office on monday, will look over it then

**Pull request merged**